crates/services/executor/src/executor.rs

```diff
@@ -565,11 +577,15 @@ where
         let block_gas_limit = self.consensus_params.block_gas_limit();
 
         let mut remaining_gas_limit = block_gas_limit.saturating_sub(data.used_gas);
-        // TODO: Handle `remaining_tx_count` https://github.com/FuelLabs/fuel-core/issues/2114
-        let remaining_tx_count = u16::MAX;
         // TODO: Handle `remaining_size` https://github.com/FuelLabs/fuel-core/issues/2133
         let remaining_size = u32::MAX;
 
+        // We allow at most u16::MAX transactions in a block, including the mint transaction.
+        // When processing l2 transactions, we must take into account transactions from the l1
+        // that have been included in the block already (stored in `data.tx_count`), as well
+        // as the final mint transaction.
+        let mut remaining_tx_count = MAX_TX_COUNT.saturating_sub(data.tx_count);
+
         let mut regular_tx_iter = l2_tx_source
```

Comment:
If the `next` returns more data than you requested, you will not catch a situation in which you have more transitions than allowed. Unit test is passing since `OnceTransactionsSource` handles it. But in the case of another `TxSource` it may not work.

Reply:
Good point. All TxSources implement next to never return more than `remaining_tx_count` transactions, but I have added a `.take` to avoid this scenario: 

Two remarks: 
1. The new behaviour of the function means that if we use a "bad" transaction source, transactions over the limit will be removed and not be included in a block. If we want to be completely safe we could have a new `SafeTransactionSource` trait with the same signature of `TransactionsSource`, and implement a struct `CacheableTransactionSource<TxSource>` which is implemented whenever `TxSource: TransactionsSource`, 
and keeps transactions over the limit in memory. 
Given that we control the implementations of `TransactionsSource`, this is probably an overkill.

2. The new change is not tested, I'd need to implement a `BadTransactionSource` and modify the commit_without_execute function to run with that transaction source. I can implement that if necessary,  but I'd prefer to open a dedicated PR for it.

Reply:
Yeah, `BadTransactionSource` sounds good=D